File: Assets/ProjectGame/Lib/data/enum/GameObject/GameObjectID.py
from enum import IntEnum
from typing import Callable, List, Any


class GameObjectID(IntEnum):
    NONE = 0  # デフォルト値（C#のNoneに相当）
    Character_kuzuha = 1  # 葛葉
    Character_Ange = 2  # アンジュ
    Character_Ryushen = 3  # 緑仙
    Character_Belmond = 4  # ベルモンド
    Character_Himawari = 5  # 本間ひわまり
    Character_Mashiro = 6  # ましろ
    MAX = 7

def to_int(id: "GameObjectID") -> int:
    return int(id)


def to_GameObjectID(id: int) -> "GameObjectID":
    return GameObjectID(id)


def to_index(id: "GameObjectID") -> int:
    return int(id) - 1


def for_id(action: Callable[["GameObjectID"], None]):
    if action is None:
        raise ValueError("action cannot be None")
    start = GameObjectID.Character_kuzuha.value
    for i in range(start, GameObjectID.MAX.value):
        try:
            value = GameObjectID(i)
            action(value)
        except ValueError:
            continue  # 未定義の値はスキップ


def find_all(predicate: Callable[["GameObjectID"], bool]) -> List["GameObjectID"]:
    if predicate is None:
        raise ValueError("predicate cannot be None")
    results: List["GameObjectID"] = []
    start = GameObjectID.Character_kuzuha.value
    for i in range(start, GameObjectID.MAX.value):
        try:
            value = GameObjectID(i)
            if predicate(value):
                results.append(value)
        except ValueError:
            continue
    return results


def find(predicate: Callable[["GameObjectID"], bool]) -> "GameObjectID":
    if predicate is None:
        raise ValueError("predicate cannot be None")
    start = GameObjectID.Character_kuzuha.value
    for i in range(start, GameObjectID.MAX.value):
        try:
            value = GameObjectID(i)
            if predicate(value):
                return value
        except ValueError:
            continue
    return GameObjectID.NONE

File: Assets/ProjectGame/Lib/data/enum/GameObject/GameObjectID.js
// GameObjectID - Pure JavaScript Version
export const GameObjectID = {
    None: 0,  // デフォルト値（C#互換）
    Character_kuzuha: 1,  // 葛葉
    Character_Ange: 2,  // アンジュ
    Character_Ryushen: 3,  // 緑仙
    Character_Belmond: 4,  // ベルモンド
    Character_Himawari: 5,  // 本間ひわまり
    Character_Mashiro: 6,  // ましろ
    Max: 7
};

export const GameObjectIDExtensions = {
    /**
     * Enumを数値に変換
     */
    toInt(id) {
        return Number(id);
    },

    /**
     * 数値をEnumに変換
     */
    toGameObjectID(id) {
        return id;
    },

    /**
     * 0-based indexに変換
     */
    toIndex(id) {
        return Number(id) - 1;
    },

    /**
     * すべてのIDに対して処理を実行
     */
    forID(action) {
        if (typeof action !== 'function') {
            throw new Error('action must be a function');
        }
        const start = GameObjectID.Character_kuzuha;
        const max = GameObjectID.Max;
        for (let id = start; id < max; id++) {
            if (Object.values(GameObjectID).includes(id)) {
                action(id);
            }
        }
    },

    /**
     * 条件に合うすべてのIDを返す
     */
    findAll(predicate) {
        if (typeof predicate !== 'function') {
            throw new Error('predicate must be a function');
        }
        const results = [];
        const start = GameObjectID.Character_kuzuha;
        const max = GameObjectID.Max;
        for (let id = start; id < max; id++) {
            if (Object.values(GameObjectID).includes(id)) {
                if (predicate(id)) {
                    results.push(id);
                }
            }
        }
        return results;
    },

    /**
     * 条件に合う最初のIDを返す（見つからなければ None）
     */
    find(predicate) {
        if (typeof predicate !== 'function') {
            throw new Error('predicate must be a function');
        }
        const start = GameObjectID.Character_kuzuha;
        const max = GameObjectID.Max;
        for (let id = start; id < max; id++) {
            if (Object.values(GameObjectID).includes(id)) {
                if (predicate(id)) {
                    return id;
                }
            }
        }
        return GameObjectID.None;
    }
};
